File: src/pia.c
// The MIT License (MIT)
//
// Copyright (c) 2015 Stefan Arentz - http://github.com/st3fan/ewm
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

#include "cpu.h"
#include "pia.h"

// This implements a 6820 Peripheral I/O Adapter. On the Apple I this
// is what connects the keyboard and display logic to the CPU. The
// implementation is not complete but does enough to support how the
// keyboard and display are hooked up.

static uint8_t pia_read(struct cpu_t *cpu, struct mem_t *mem, uint16_t addr) {
   struct ewm_pia_t *pia = (struct ewm_pia_t*) mem->obj;
   switch (addr) {
      case EWM_A1_PIA6820_KBD_DDR:
         if (pia->ctla & 0b00000100) {
            pia->ctla &= 0b01111111; // Clear IRQA1
            return (pia->outa & pia->ddra) | (pia->ina & ~pia->ddra);
         } else {
            return pia->ddra;
         }
         break;
      case EWM_A1_PIA6820_KBD_CTL:
         return pia->ctla;
         break;
      case EWM_A1_PIA6820_DSP_DDR:
         if (pia->ctlb & 0b00000100) {
            return (pia->outb & pia->ddrb) | (pia->inb & ~pia->ddrb);
         } else {
            return pia->ddrb;
         }
         break;
      case EWM_A1_PIA6820_DSP_CTL:
         return pia->ctlb;
         break;
   }
   return 0;
}

static void pia_write(struct cpu_t *cpu, struct mem_t *mem, uint16_t addr, uint8_t v) {
   struct ewm_pia_t *pia = (struct ewm_pia_t*) mem->obj;
   switch (addr) {
      case EWM_A1_PIA6820_KBD_DDR:
         // Check B2 (DDR Access)
         if (pia->ctla & 0b00000100) {
            // Write output register and run callback
            pia->outa = v;
            if (pia->callback) {
               pia->callback(pia, pia->callback_obj, EWM_PIA6820_DDRA, v);
            }
         } else {
            // Write DDR register
            pia->ddra = v;
            // TODO Do we need a callback? Not relevant for Apple 1?
         }
         break;
      case EWM_A1_PIA6820_KBD_CTL:
         pia->ctla = (v & 0b00111111);
         break;
      case EWM_A1_PIA6820_DSP_DDR:
         // Check B2 (DDR Access)
         if (pia->ctlb & 0b00000100) {
            // Write output register and run callback
            pia->outb = v;
            if (pia->callback) {
               pia->callback(pia, pia->callback_obj, EWM_PIA6820_DDRB, v);
            }
         } else {
            // Write DDR register
            pia->ddrb = v;
            // TODO Do we need a callback? Not relevant for Apple 1?
         }
         break;
      case EWM_A1_PIA6820_DSP_CTL:
         pia->ctlb = (v & 0b00111111);
         break;
   }
}

static int ewm_pia_init(struct ewm_pia_t *pia, struct cpu_t *cpu) {
   memset(pia, 0, sizeof(struct ewm_pia_t));
   cpu_add_iom(cpu, EWM_A1_PIA6820_ADDR, EWM_A1_PIA6820_ADDR + EWM_A1_PIA6820_LENGTH - 1, pia, pia_read, pia_write);
   return 0;
}

struct ewm_pia_t *ewm_pia_create(struct cpu_t *cpu) {
   struct ewm_pia_t *pia = (struct ewm_pia_t*) malloc(sizeof(struct ewm_pia_t));
   if (ewm_pia_init(pia, cpu) != 0) {
      free(pia);
      pia = NULL;
   }
   return pia;
}

void ewm_pia_destroy(struct ewm_pia_t *pia) {
   free(pia);
}

void ewm_pia_set_outa(struct ewm_pia_t *pia, uint8_t v) {
   pia->outa = v;
}

void ewm_pia_set_ina(struct ewm_pia_t *pia, uint8_t v) {
   pia->ina = v;
}

void ewm_pia_set_outb(struct ewm_pia_t *pia, uint8_t v) {
   pia->outb = v;
}

void ewm_pia_set_inb(struct ewm_pia_t *pia, uint8_t v) {
   pia->inb = v;
}

void ewm_pia_set_irqa1(struct ewm_pia_t *pia) {
   pia->ctla |= 0b10000000; // Set IRQA1
}

File: ewm-core/src/pia.rs
//! 6820 Peripheral I/O Adapter, port of `pia.c`. On the Apple 1 this is
//! what connects the keyboard and display logic to the CPU. The
//! implementation is not complete but does enough to support how the
//! keyboard and display are hooked up. As in C the PIA registers itself as
//! an IO region; the output callback becomes a drainable queue.

use crate::mem::Device;

pub const PIA6820_DDRA: u8 = 0;
pub const PIA6820_CTLA: u8 = 1;
pub const PIA6820_DDRB: u8 = 2;
pub const PIA6820_CTLB: u8 = 3;

pub const A1_PIA6820_ADDR: u16 = 0xd010;
pub const A1_PIA6820_LENGTH: u16 = 0x0100;

const KBD_DDR: u16 = A1_PIA6820_ADDR + PIA6820_DDRA as u16;
const KBD_CTL: u16 = A1_PIA6820_ADDR + PIA6820_CTLA as u16;
const DSP_DDR: u16 = A1_PIA6820_ADDR + PIA6820_DDRB as u16;
const DSP_CTL: u16 = A1_PIA6820_ADDR + PIA6820_CTLB as u16;

#[derive(Default)]
pub struct Pia {
    pub ina: u8,
    pub outa: u8,
    pub ddra: u8,
    pub ctla: u8,
    pub inb: u8,
    pub outb: u8,
    pub ddrb: u8,
    pub ctlb: u8,
    out: Vec<(u8, u8)>,
}

impl Pia {
    pub fn new() -> Pia {
        Pia::default()
    }

    pub fn set_ina(&mut self, v: u8) {
        self.ina = v;
    }

    pub fn set_irqa1(&mut self) {
        self.ctla |= 0b1000_0000; // Set IRQA1
    }

    /// `(ddr, v)` pairs written to the output registers since the last
    /// drain — the C output callback turned into a queue for the machine to
    /// route to its display sink.
    pub fn drain_out(&mut self) -> Vec<(u8, u8)> {
        std::mem::take(&mut self.out)
    }
}

impl Device for Pia {
    /// Port of `pia_read`. Reading the keyboard register clears IRQA1.
    fn read(&mut self, addr: u16, _cycles: u64) -> u8 {
        match addr {
            KBD_DDR => {
                if self.ctla & 0b0000_0100 != 0 {
                    self.ctla &= 0b0111_1111; // Clear IRQA1
                    (self.outa & self.ddra) | (self.ina & !self.ddra)
                } else {
                    self.ddra
                }
            }
            KBD_CTL => self.ctla,
            DSP_DDR => {
                if self.ctlb & 0b0000_0100 != 0 {
                    (self.outb & self.ddrb) | (self.inb & !self.ddrb)
                } else {
                    self.ddrb
                }
            }
            DSP_CTL => self.ctlb,
            _ => 0,
        }
    }

    /// Port of `pia_write`. Output-register writes are queued for
    /// `drain_out`, where the C invoked the callback.
    fn write(&mut self, addr: u16, v: u8, _cycles: u64) {
        match addr {
            KBD_DDR => {
                // Check B2 (DDR Access)
                if self.ctla & 0b0000_0100 != 0 {
                    self.outa = v;
                    self.out.push((PIA6820_DDRA, v));
                } else {
                    self.ddra = v;
                }
            }
            KBD_CTL => self.ctla = v & 0b0011_1111,
            DSP_DDR => {
                // Check B2 (DDR Access)
                if self.ctlb & 0b0000_0100 != 0 {
                    self.outb = v;
                    self.out.push((PIA6820_DDRB, v));
                } else {
                    self.ddrb = v;
                }
            }
            DSP_CTL => self.ctlb = v & 0b0011_1111,
            _ => {}
        }
    }
}
